src/protocol/codecs/algomsg.rs

```diff
@@ -0,0 +1,155 @@
+use std::io;
+
+use bytes::BytesMut;
+use tokio_util::codec::Decoder;
+use tracing::*;
+
+use super::topic::TopicCodec;
+use crate::protocol::invalid_data;
+
+/// [Tag] represents a message type identifier.
+///
+/// The original tag list can be found in go-algorand/protocol/tags.go.
+#[derive(Debug)]
+pub enum Tag {
+    UnknownMsg,
+    AgreementVote,
+    MsgOfInterest,
+    MsgDigestSkip,
+    NetPrioResponse,
+    Ping,
+    PingReply,
+    ProposalPayload,
+    StateProofSig,
+    TopicMsgResp,
+    Txn,
+    UniCatchupReq,
+    UniEnsBlockReq,
+    VoteBundle,
+}
+
+impl TryFrom<BytesMut> for Tag {
+    type Error = io::Error;
+
+    fn try_from(tag: BytesMut) -> Result<Self, Self::Error> {
+        let tag = std::str::from_utf8(&tag)
+            .map_err(|_| invalid_data!("couldn't convert the tag to a utf8 string"))?;
+
+        Ok(match tag {
+            "??" => Self::UnknownMsg,
+            "AV" => Self::AgreementVote,
+            "MI" => Self::MsgOfInterest,
+            "MS" => Self::MsgDigestSkip,
+            "NP" => Self::NetPrioResponse,
+            "pi" => Self::Ping,
+            "pj" => Self::PingReply,
+            "PP" => Self::ProposalPayload,
+            "SP" => Self::StateProofSig,
+            "TS" => Self::TopicMsgResp,
+            "TX" => Self::Txn,
+            "UC" => Self::UniCatchupReq,
+            "UE" => Self::UniEnsBlockReq,
+            "VB" => Self::VoteBundle,
+            _ => return Err(invalid_data!("unexpected tag")),
+        })
+    }
+}
+
+#[derive(Debug)]
+pub enum Payload {
+    MsgOfInterest(MsgOfInterest),
+    NotImplemented,
+}
+
+/// [MsgOfInterest] contains a tag list in which the node is interested.
+#[derive(Debug)]
+pub struct MsgOfInterest {
+    /// Message tags for which the node is interested (subscribed).
+    // TODO(Rqnsom): convert to a hashset.
+    pub tags: String,
+}
+
+impl MsgOfInterest {
+    /// Decodes a byte stream from the [input] using the [topic] codec to generate a [MsgOfInterest] message.
+    fn from(input: &mut BytesMut, topic: &mut TopicCodec) -> Result<Self, io::Error> {
+        let mut topics = topic
+            .decode(input)?
+            .ok_or_else(|| invalid_data!("couldn't decode the topics"))?;
+        if topics.len() != 1 {
+            return Err(invalid_data!("expected a single topic"));
+        }
+
+        let tag_topic = topics.pop().expect("couldn't pop the 'tags' topic");
```

Comment:
```suggestion
        let tag_topic = topics.pop().unwrap();
```
guaranteed to work by L78